numpy/_core/src/multiarray/public_dtype_api.c

```diff
@@ -0,0 +1,180 @@
+            PyErr_SetString(PyExc_RuntimeError,
+                    "Parametric DType must define a common-instance and "
+                    "descriptor discovery function!");
+            return -1;
+        }
+    }
+
+    if (NPY_DT_SLOTS(DType)->within_dtype_castingimpl == NULL) {
+        /*
+         * We expect this for now. We should have a default for DType that
+         * only supports simple copy (and possibly byte-order assuming that
+         * they swap the full itemsize).
+         */
+        PyErr_SetString(PyExc_RuntimeError,
+                "DType must provide a function to cast (or just copy) between "
+                "its own instances!");
+        return -1;
+    }
+
+    /* And finally, we have to register all the casts! */
+    return 0;
+}
+
+
+NPY_NO_EXPORT PyObject *
+_get_dtype_api(PyObject *NPY_UNUSED(mod), PyObject *NPY_UNUSED(arg))
+{
+    static void *api_table[40] = {0};
+
+    /* The type of the DType metaclass */
+    api_table[0] = &PyArrayDTypeMeta_Type;
+    /* Boolean */
+    api_table[1] = &PyArray_BoolDType;
+    /* Integers */
+    api_table[2] = &PyArray_ByteDType;
+    api_table[3] = &PyArray_UByteDType;
+    api_table[4] = &PyArray_ShortDType;
+    api_table[5] = &PyArray_UShortDType;
+    api_table[6] = &PyArray_IntDType;
+    api_table[7] = &PyArray_UIntDType;
+    api_table[8] = &PyArray_LongDType;
+    api_table[9] = &PyArray_ULongDType;
+    api_table[10] = &PyArray_LongLongDType;
+    api_table[11] = &PyArray_ULongLongDType;
+    /* Integer aliases */
+    api_table[12] = &PyArray_Int8DType;
+    api_table[13] = &PyArray_UInt8DType;
+    api_table[14] = &PyArray_Int16DType;
+    api_table[15] = &PyArray_UInt16DType;
+    api_table[16] = &PyArray_Int32DType;
+    api_table[17] = &PyArray_UInt32DType;
+    api_table[18] = &PyArray_Int64DType;
+    api_table[19] = &PyArray_UInt64DType;
+    api_table[20] = &PyArray_IntpDType;
+    api_table[21] = &PyArray_UIntpDType;
+    /* Floats */
+    api_table[22] = &PyArray_HalfDType;
+    api_table[23] = &PyArray_FloatDType;
+    api_table[24] = &PyArray_DoubleDType;
+    api_table[25] = &PyArray_LongDoubleDType;
+    /* Complex */
+    api_table[26] = &PyArray_CFloatDType;
+    api_table[27] = &PyArray_CDoubleDType;
+    api_table[28] = &PyArray_CLongDoubleDType;
+    /* String/Bytes */
+    api_table[29] = &PyArray_BytesDType;
+    api_table[30] = &PyArray_UnicodeDType;
+    /* Datetime/Timedelta */
+    api_table[31] = &PyArray_DatetimeDType;
+    api_table[32] = &PyArray_TimedeltaDType;
+    /* Object and Structured */
+    api_table[33] = &PyArray_ObjectDType;
+    api_table[34] = &PyArray_VoidDType;
+    /* Abstract */
+    api_table[35] = &PyArray_PyIntAbstractDType;
+    api_table[36] = &PyArray_PyFloatAbstractDType;
+    api_table[37] = &PyArray_PyComplexAbstractDType;
+    api_table[38] = &PyArray_DefaultIntDType;
+    /* Non-legacy DTypes that are built in to NumPy */
+    api_table[39] = &PyArray_StringDType;
```

Comment:
So I think I had suggested that before, but just mentionig it and hoping @mattip can voice an opinion:

An alternative that I rather like would be to add a big chunk of "empty slots" with a common to the api-gen (i.e. just with a `+ list(range(...))`) and then fill the normal API table here and above with that offset.

There isn't a huge reason for it, but I somehwat like the idea of having fewer structs.

---

I may have some tweaks if I review the API, but I itself, but I would prefer to just push this through, and the docs look nice on a quick glance.
(Tweaking this API is completely fine also after RC, the 0 people using it in their wheels already then, can recompile :))